### ISSUE TITLE
Inconsistent dimension when generate contact maps and maximum sequence length

### ISSUE THREAD

**Issue by Douglas2Code** [facebookresearch/esm#34]
First of all, thank you for your fantastic work! I tried your embeddings on some basic protein family tasks and the results are amazing!

However, I have some issues when I try to generate contact map predictions. 1. From your code, the batch_conventer function will padding sequences to the maximum length. In my case is 600. Whereas, the generated contact maps have the dimension of [batch_size, 599 or 600, 599 or 600]. I am wondering why this happens. 2. Another issue is with the maximum length of the sequence the model can proceed, I am working with protein families that can have a maximum of over 30,000 long. And when I trying to use the model on such protein, the position encoding seems to fail. So I am curious what is the maximum length of protein this model can handle so I can set a reasonable cutoff.

**Comment by joshim5** [comment]
Hi @Douglas2Code, thanks for your interest and glad that your results with ESM are looking so good!

To answer your questions:
1. You may have to crop the output of the generated contact maps if the input contains padding tokens. If you provide a code example, we could give a more detailed answer of how to do this.
2. The model was only trained with sequences <1024 tokens. We randomly cropped longer sequences (different crop at each epoch). Since the model hasn't been trained on longer sequences, it's not possible to make predictions beyond that length, as the positional encodings have not been trained. We haven't experimented with sequences >1024 in length, but we have some ideas of how the model could be applied. However, we can't provide any guarantees whether you'll get good performance, so I would recommend using sequences <1024 length if possible. See #21 for a few ideas.

**Comment by nasserhashemi** [comment]
@Douglas2Code  maybe this work can help you to deal with loner sequence as well:) 
https://arxiv.org/pdf/2012.03084.pdf
https://arxiv.org/abs/2004.05150

**Comment by Douglas2Code** [comment]
> Hi @Douglas2Code, thanks for your interest and glad that your results with ESM are looking so good!
> 
> To answer your questions:
> 
> 1. You may have to crop the output of the generated contact maps if the input contains padding tokens. If you provide a code example, we could give a more detailed answer of how to do this.
> 2. The model was only trained with sequences <1024 tokens. We randomly cropped longer sequences (different crop at each epoch). Since the model hasn't been trained on longer sequences, it's not possible to make predictions beyond that length, as the positional encodings have not been trained. We haven't experimented with sequences >1024 in length, but we have some ideas of how the model could be applied. However, we can't provide any guarantees whether you'll get good performance, so I would recommend using sequences <1024 length if possible. See #21 for a few ideas.

Thank you for your reply!

I actually have one extra question regarding sequences that have been padded. From the contact maps, it looks like the extra padding positions are filter out but at the end of the sequences there would be very high contact maps activations. I am wondering why this happens?

![CMGC](https://user-images.githubusercontent.com/48750976/107460619-6c9ba100-6b26-11eb-8f66-8ba8625a15ca.png)

**Comment by rmrao** [comment]
Hi @Douglas2Code, it looks like there's an issue with the end of sentence token being incorrectly filtered out. Unfortunately, I'm not able to reproduce this error locally (here's an example with three sequences of different length - the eos token appears to be consistently filtered out).
<img width="996" alt="image" src="https://user-images.githubusercontent.com/6496605/107524759-fd946b80-6b6a-11eb-82f7-08bc86502d1b.png">

If you could provide a minimum working example, or the sequences you are using when this error occurs, that might help reproduce the issue.

**Comment by Douglas2Code** [comment]
[sample.txt](https://github.com/facebookresearch/esm/files/5960552/sample.txt)
Thank you for your reply! Here are some example sequences that I used. They were input as a batch with a maximum length set as 600. 
![AAK1](https://user-images.githubusercontent.com/48750976/107554842-78c64380-6ba4-11eb-82f7-f24fa4c44544.png)
Here is one output. 

The code I used for generating contact maps:

# this function is for generate contact maps using esm-1b model layer33
def generate_contact_test(dest_dir, model, batch_labels, batch_strs, batch_tokens, target_layer=33):
    results = []
    with torch.no_grad():
        print(batch_tokens.shape)
        out = model(batch_tokens, repr_layers=[target_layer], return_contacts=True)
        results.append(out["contacts"])
        torch.save(out["contacts"], dest_dir+'/'+'contact'+str(batch_labels)+'.pt')
        gc.collect()
    return results

**Comment by rmrao** [comment]
It looks like the issue is with the ordering of certain operations. The following workflow should work:

```python
data = [("protein1", "..."), ("protein2", "..."), ...]
batch_size = 8
results = []
for i in range(0, len(data), batch_size):
    batch = data[i:i + batch_size]
    batch_labels, batch_strs, batch_tokens = batch_converter(batch)
    contacts = model.predict_contacts(batch_tokens)
    results.append(contacts)
```

Note in particular - you should do the batching first and run batch converter only on the subset you plan to pass to the model, rather than converting all your data with the batch converter and then batching the tokens. This will be more efficient anyway (since the batch converter will only add as many padding tokens as necessary). We will also shortly push a change that fixes this even if you convert first.

**Comment by Douglas2Code** [comment]
Got it! Thank you for helping out!
